redwood-tooling-schema/src/main/kotlin/app/cash/redwood/tooling/schema/schemaJson.kt

```diff
@@ -0,0 +1,272 @@
+import kotlin.reflect.KTypeProjection
+import kotlin.reflect.KVariance.IN
+import kotlin.reflect.KVariance.INVARIANT
+import kotlin.reflect.KVariance.OUT
+import kotlinx.serialization.EncodeDefault
+import kotlinx.serialization.EncodeDefault.Mode.ALWAYS
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+
+public data class EmbeddedSchema(
+  val path: String,
+  val json: String,
+)
+
+/**
+ * Convert this schema to JSON which can be embedded inside the schema artifact.
+ * This JSON will be read when the schema is used as a dependency.
+ */
+public fun ProtocolSchema.toEmbeddedSchema(): EmbeddedSchema {
+  val schemaJson = JsonSchemaV1(
+    type = JsonType(listOf(`package`, name)),
+    scopes = scopes.map { it.toJsonType() },
+    widgets = widgets.map { widget ->
+      JsonWidget(
+        type = widget.type.toJsonType(),
+        traits = widget.traits.map { trait ->
+          when (trait) {
+            is ProtocolProperty -> {
+              JsonProperty(
+                tag = trait.tag,
+                name = trait.name,
+                type = trait.type.toJsonType(),
+                defaultExpression = trait.defaultExpression,
+              )
+            }
+            is ProtocolEvent -> {
+              JsonEvent(
+                tag = trait.tag,
+                name = trait.name,
+                parameterType = trait.parameterType?.toJsonType(),
+                defaultExpression = trait.defaultExpression,
+              )
+            }
+            is ProtocolChildren -> {
+              JsonChildren(
+                tag = trait.tag,
+                name = trait.name,
+                scope = trait.scope?.toJsonType(),
+              )
+            }
+          }
+        },
+      )
+    },
+    layoutModifiers = layoutModifiers.map { layoutModifier ->
+      JsonLayoutModifier(
+        tag = layoutModifier.tag,
+        type = layoutModifier.type.toJsonType(),
+        scopes = layoutModifier.scopes.map { it.toJsonType() },
+        properties = layoutModifier.properties.map {
+          JsonLayoutModifierProperty(
+            name = it.name,
+            type = it.type.toJsonType(),
+            defaultExpression = it.defaultExpression,
+          )
+        }
+      )
+    },
+    dependencies = dependencies.map {
+      JsonType(listOf(it.`package`, it.name))
+    },
+  )
+  val json = json.encodeToString(JsonSchemaV1.serializer(), schemaJson)
+
+  return EmbeddedSchema(
+    path = "${`package`.replace('.', '/')}/$name.json",
+    json = json,
+  )
+}
+
+private val json = Json {
+  prettyPrint = true
+  prettyPrintIndent = "\t"
+  classDiscriminator = "kind"
+}
+
+@Serializable
+private data class JsonType(
+  val names: List<String>,
+  val variance: Variance = Invariant,
+  val parameterTypes: List<JsonType> = emptyList(),
+  val nullable: Boolean = false,
```

Comment:
This can represent 95% of Kotlin types, and 100% of the types we care about. We do not support type variables on widgets, so they don't need to show up in our types.

This type will have to move into the public API soon, and I'll test it then.